### DISCUSSION TITLE
Differences between mjs and elk

### DISCUSSION THREAD

**Discussion by maelp** [cesanta/elk#44] [Q&A] [answered]
Hi,

thanks for those amazing embedded js projects, I'm curious since there is both `mjs` and `elk`, what would be the differences between them, and when should one or the other be used? (in my case, this would be to use it in an ESP32 setting, so there is some memory available, perhaps mjs is better suited?)

**Comment by cpq** [comment, marked as answer]
Elk is way smaller, simple, and bare-bones. For any project, we recommend Elk - it provides a greater control for your JS environment. The exact list of differences - well, just traverse READMEs for both projects and compare the API.

In fact, we have one more JS engine - https://github.com/cesanta/v7. That one is even more complete and complex than mjs. And no, we won't go and describe all bits & pieces of differences between those or other engines. The bigger & more complex the engine is, the more feature-complete it is. For ESP32 or other embedded MCU, use Elk. Unsure? Do your research.

**Reply by maelp** [reply]
Ok, I saw v7 but it seemed to be deprecated in favor of mjs, I guess mjs is a bit more feature-complete then

**Reply by cpq** [reply]
Yeah, mjs is more feature complete. It has arrays, for example. Elk does not. The intention behind Elk is that everything that needs to be in the JS environment, should be exported from C. JS just orchestrates things. Need a fancy string manipulation functionality? Do it in C, export to JS. Need some fancy peripheral support ? Again, C -> JS. This way things will remain fast, and scriptable by the end-user.

We use MCUs and ESP32 in particular a lot. Our engine of choice is Elk.

**Reply by cpq** [reply]
Another important thing to keep in mind is side-effects of the exported C functions. There is a issue which does not depend on engine.
For example, you've imported a C function `set_timer()` which allocates a C timer - the side-effect is an allocated thing in C land which JS has no access to. Of course a `clear_timer()` should also be imported, but there is no guarantee that JS user calls it - especially when the engine gets destroyed for some reason (e.g. JS script gets replaced). To deal with it the best way, a function that is imported by JS should not be `set_timer()`, but a wrapper around `set_timer` that keeps track of the allocation, and cleans up in case JS user did not call `clear_timer()`.

Again, all this is regardless of the engine used.

**Reply by maelp** [reply]
Thanks!

Speaking of garbage collection, is that done automatically for JS objects?

For instance if I define a function like

```
    jsval_t c_getState()
    {
        jsval_t state = js_mkobj(js);
        js_set(js, state, "someSensor", js_mknum(js, 16.21));
        js_set(js, state, "someValue", js_mknum(js, 10));
        return state;
    }
```

and that I call this code repeatedly from the JS code, are old objects released when they get out of scope?
When is the GC run?

**Reply by maelp** [reply]
For instance the JS code would call in a loop:

```js
timer(1000, function() {
   state = getState();
   if (state.someSensor >= 16) { do_something(); }
});
```

**Reply by cpq** [reply]
Yes, JS is done automatically before each top-level statement if the free memory exceeds a threshold.
Currently, threshold is set to 75% of the available memory.
